From patchwork Tue Feb 19 03:53:45 2019
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Suraj Jitindar Singh <sjitindarsingh@gmail.com>
X-Patchwork-Id: 10819167
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id E3D6D1399
	for <patchwork-kvm@patchwork.kernel.org>;
 Tue, 19 Feb 2019 03:54:18 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id C66AE28816
	for <patchwork-kvm@patchwork.kernel.org>;
 Tue, 19 Feb 2019 03:54:18 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id B972028965; Tue, 19 Feb 2019 03:54:18 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-8.0 required=2.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,FREEMAIL_FROM,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI
	autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 55B9528816
	for <patchwork-kvm@patchwork.kernel.org>;
 Tue, 19 Feb 2019 03:54:18 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1725768AbfBSDyH (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Mon, 18 Feb 2019 22:54:07 -0500
Received: from mail-pl1-f194.google.com ([209.85.214.194]:44821 "EHLO
        mail-pl1-f194.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1725497AbfBSDyH (ORCPT <rfc822;kvm@vger.kernel.org>);
        Mon, 18 Feb 2019 22:54:07 -0500
Received: by mail-pl1-f194.google.com with SMTP id c4so4680130pls.11;
        Mon, 18 Feb 2019 19:54:06 -0800 (PST)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=gmail.com; s=20161025;
        h=from:to:cc:subject:date:message-id;
        bh=Hpae+Cn7fvx8SZf70hgI8zVNXSz5QD/XB5kgI4NdgAs=;
        b=s99RJ+VMFCimGO1Nwnbfa5/KjipGbUFNQaZyFlkTttbj35BOZRWiYsFPv3o6fYpEpW
         1e60kKMPcQkD0hxbgZJzthOgnLveMrP/xui76kVxgRxB4wyjjselwGvO7r422o5Vf/4A
         wPoxKC8E4hGW3FAmdRnDflOBWa1r7mOctRA7SI/kVBRzhRI7go8ou5fkCf86Sr99J71u
         rW8pavpLb3jnybR2PY6+4Ghd/PeL7VxFvTzQN5RgQ1/9fRpBNICmVSOfuW4liu9q+xrN
         VxqGo5uN8ylvKpHX4oy+tX445LY+rxNXoWm4lShiWtjj5HZXqwkBuaklm0w0I5qNGA5E
         yQ/Q==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:from:to:cc:subject:date:message-id;
        bh=Hpae+Cn7fvx8SZf70hgI8zVNXSz5QD/XB5kgI4NdgAs=;
        b=GXqlTJq2ONDPbyQQX9G8BTjo87aiPtZqKe0o2VzTampP1Z277oemPDakQsaaLHr7Cq
         yij37PHX7chz/ZwE4Mbz8Uj7pZqcPDt1mFdiILbI5kEe4fB7Jhhpori99v6wafU16m5r
         guixIhVn6nHfRwBwmGPK6nuMRIMeir2A6m2nhoNwj24S34XjgozxB0X44f7qb0tyceYZ
         7a+sxlxAjZZHdteUn/EYL8pI+0uXFQzusazUVyxwR0jJuGGQwXuEALNiG1ilvk49BqZ9
         65kXESuQ/rYsFOX77IdhgIBD0bBkcBnWlKR9Ch7RLLfXAtdkeAXOyWTnca/FkIgQNIZF
         66PA==
X-Gm-Message-State: AHQUAuYCoDnr3KrKy1m57olwAmX5bP8oHg9M+YGJy6/8EvRIQI6WYVel
        40xBT8aH49udg7WeVB48rXiYy69s
X-Google-Smtp-Source: 
 AHgI3IbCEMpS0yt5WXYpWM2xJnHePHC6Ak+K0uqB0n9yH7/9knXMN6whRl4qiyRTpLwvcVXlyQAvPQ==
X-Received: by 2002:a17:902:4827:: with SMTP id
 s36mr28148457pld.168.1550548446349;
        Mon, 18 Feb 2019 19:54:06 -0800 (PST)
Received: from surajjs2.ozlabs.ibm.com ([122.99.82.10])
        by smtp.gmail.com with ESMTPSA id
 o6sm14116536pgo.27.2019.02.18.19.54.03
        (version=TLS1_2 cipher=ECDHE-RSA-CHACHA20-POLY1305 bits=256/256);
        Mon, 18 Feb 2019 19:54:05 -0800 (PST)
From: Suraj Jitindar Singh <sjitindarsingh@gmail.com>
To: kvm-ppc@vger.kernel.org
Cc: kvm@vger.kernel.org, linuxppc-dev@lists.ozlabs.org,
        paulus@ozlabs.org, Suraj Jitindar Singh <sjitindarsingh@gmail.com>
Subject: [PATCH v2] KVM: PPC: Book3S: Add KVM stat largepages_[2M/1G]
Date: Tue, 19 Feb 2019 14:53:45 +1100
Message-Id: <20190219035345.26239-1-sjitindarsingh@gmail.com>
X-Mailer: git-send-email 2.13.6
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

This adds an entry to the kvm_stats_debugfs directory which provides the
number of large (2M or 1G) pages which have been used to setup the guest
mappings.

Signed-off-by: Suraj Jitindar Singh <sjitindarsingh@gmail.com>
---

V1 -> V2:
- Rename debugfs files from num_[2M/1G]_pages to largepages_[2M/1G] to match
  x86

 arch/powerpc/include/asm/kvm_host.h    |  2 ++
 arch/powerpc/kvm/book3s.c              |  3 +++
 arch/powerpc/kvm/book3s_64_mmu_radix.c | 15 ++++++++++++++-
 3 files changed, 19 insertions(+), 1 deletion(-)

diff --git a/arch/powerpc/include/asm/kvm_host.h b/arch/powerpc/include/asm/kvm_host.h
index 0f98f00da2ea..cbb090010312 100644
--- a/arch/powerpc/include/asm/kvm_host.h
+++ b/arch/powerpc/include/asm/kvm_host.h
@@ -99,6 +99,8 @@ struct kvm_nested_guest;
 
 struct kvm_vm_stat {
 	ulong remote_tlb_flush;
+	ulong num_2M_pages;
+	ulong num_1G_pages;
 };
 
 struct kvm_vcpu_stat {
diff --git a/arch/powerpc/kvm/book3s.c b/arch/powerpc/kvm/book3s.c
index bd1a677dd9e4..72fd7d44379b 100644
--- a/arch/powerpc/kvm/book3s.c
+++ b/arch/powerpc/kvm/book3s.c
@@ -39,6 +39,7 @@
 #include "book3s.h"
 #include "trace.h"
 
+#define VM_STAT(x) offsetof(struct kvm, stat.x), KVM_STAT_VM
 #define VCPU_STAT(x) offsetof(struct kvm_vcpu, stat.x), KVM_STAT_VCPU
 
 /* #define EXIT_DEBUG */
@@ -71,6 +72,8 @@ struct kvm_stats_debugfs_item debugfs_entries[] = {
 	{ "pthru_all",       VCPU_STAT(pthru_all) },
 	{ "pthru_host",      VCPU_STAT(pthru_host) },
 	{ "pthru_bad_aff",   VCPU_STAT(pthru_bad_aff) },
+	{ "largepages_2M",    VM_STAT(num_2M_pages) },
+	{ "largepages_1G",    VM_STAT(num_1G_pages) },
 	{ NULL }
 };
 
diff --git a/arch/powerpc/kvm/book3s_64_mmu_radix.c b/arch/powerpc/kvm/book3s_64_mmu_radix.c
index 1b821c6efdef..f55ef071883f 100644
--- a/arch/powerpc/kvm/book3s_64_mmu_radix.c
+++ b/arch/powerpc/kvm/book3s_64_mmu_radix.c
@@ -403,8 +403,13 @@ void kvmppc_unmap_pte(struct kvm *kvm, pte_t *pte, unsigned long gpa,
 		if (!memslot)
 			return;
 	}
-	if (shift)
+	if (shift) { /* 1GB or 2MB page */
 		page_size = 1ul << shift;
+		if (shift == PMD_SHIFT)
+			kvm->stat.num_2M_pages--;
+		else if (shift == PUD_SHIFT)
+			kvm->stat.num_1G_pages--;
+	}
 
 	gpa &= ~(page_size - 1);
 	hpa = old & PTE_RPN_MASK;
@@ -878,6 +883,14 @@ int kvmppc_book3s_instantiate_page(struct kvm_vcpu *vcpu,
 		put_page(page);
 	}
 
+	/* Increment number of large pages if we (successfully) inserted one */
+	if (!ret) {
+		if (level == 1)
+			kvm->stat.num_2M_pages++;
+		else if (level == 2)
+			kvm->stat.num_1G_pages++;
+	}
+
 	return ret;
 }
 
